README.md
# 📫 Iris.vim  [![ko-fi](https://img.shields.io/badge/ko--fi-Buy%20me%20a%20coffee!-red)](https://ko-fi.com/soywod#linkModal)

Simple mail client for Vim, inspired by (Neo)Mutt and Alpine.

![image](https://user-images.githubusercontent.com/10437171/84173056-07664c00-aa7d-11ea-919f-a973120a8439.png)

## Table of contents

  - [Motivation](#motivation)
  - [Requirements](#requirements)
  - [Installation](#installation)
  - [Configuration](#configuration)
    - [Identity](#identity-required)
    - [IMAP](#imap-required)
    - [SMTP](#identity-required)
    - [Passwords](#passwords)
    - [Idle mode](#idle-mode)
    - [Pagination](#pagination)
    - [Attachments](#attachments)
  - [Usage](#usage)
    - [Email list](#list-mails)
    - [Email text preview](#email-text-preview)
    - [Email composition](#email-composition)
    - [Folders](#folders)
    - [Flags](#flags)
    - [Contacts](#contacts)
  - [Changelog](https://github.com/soywod/iris.vim/blob/master/CHANGELOG.md)
  - [Contributing](#contributing)

## Motivation

(Neo)Mutt and Alpine are very good terminal mail clients, but they lack of Vim
mappings. You can emulate, but it requires a lot of time, patience and
configuration. Why trying to emulate, when you can have it in Vim? VimL and
Python are strong enough to do so. The aim of Iris is to provide a simple mail
client that:

  - Allows you to manage your mails inside Vim
  - Does not slow down neither Vim nor your workflow (async+lazy)
  - Is built on the top of a robust [Python IMAP client](https://github.com/mjs/imapclient) to avoid implementing IMAP protocol logic

## Requirements

You need either Neovim or Vim8+ with:

  - Python3 support enabled `:echo has("python3")`
  - Job enabled `:echo has("job")`
  - Channel enabled `:echo has("channel")`

## Installation

For eg. with [`vim-plug`](https://github.com/junegunn/vim-plug):

```vim
Plug "soywod/iris.vim"
```

## Configuration

Before using Iris, you need to configure it:

### Identity (required)

```vim
let g:iris_name  = "My name"
let g:iris_mail = "your@mail.com"
```

### IMAP (required)

```vim
let g:iris_imap_host  = "your.imap.host"
let g:iris_imap_port  = 993
let g:iris_imap_login = "Your IMAP login" "Default to g:iris_mail
```

### SMTP (required)

```vim
let g:iris_smtp_host  = "your.smtp.host" "Default to g:iris_imap_host
let g:iris_smtp_port  = 587
let g:iris_smtp_login = "Your IMAP login" "Default to g:iris_mail
```

### Passwords

On startup, Iris always asks for your IMAP and SMTP passwords. To avoid this,
you can save your password in a file and encrypt it via
[GPG](https://gnupg.org/):

```bash
gpg --encrypt --sign --armor --output myfile.gpg myfile
```

```vim
let g:iris_imap_passwd_filepath = "/path/to/imap.gpg"
let g:iris_smtp_passwd_filepath = "/path/to/smtp.gpg"
```

Note: `iris_imap_passwd_filepath` & `iris_smtp_password_filepath` must be absolute paths.  For example the path `~/passwords/imap.gpg` would not work because of the `~`.

If you want to use something else than GPG, you can set up your custom command.
For eg., with the MacOSX `security` tool:

```vim
let g:iris_imap_passwd_show_cmd = "security find-internet-password -gs IMAP_KEY -w"
let g:iris_smtp_passwd_show_cmd = "security find-internet-password -gs SMTP_KEY -w"
```

### Idle mode

On startup, Iris spawns two Python jobs: one for the API, one for the [idle
mode](https://imapclient.readthedocs.io/en/2.1.0/advanced.html#watching-a-mailbox-using-idle).
The last one allows you to receive notifications on new mails. You can disable
this option or change the default timeout (every 15s):

```vim
let g:iris_idle_enabled = 1
let g:iris_idle_timeout = 15
```

### Pagination

By default, Iris fetches your last 50 mails:

```vim
let g:iris_emails_chunk_size = 50
```

*Note: the pagination is based on message sequences which is not necessary
consecutive. It makes the pagination less accurate (doesn't fetch always the
same amount of mails) but more performant.*

### Attachments

```vim
let g:iris_download_dir = "~/Downloads"
```

## Usage

### Email list

```vim
:Iris
```

Function | Default keybind | Override
--- | --- | ---
Preview (text) | `<Enter>` | `nmap <cr> <plug>(iris-preview-text-email)`
Preview (html) | `gp` (for `go preview`) | `nmap gp <plug>(iris-preview-html-email)`
Download attachments | `ga` (for `go attachments`) | `nmap ga <plug>(iris-download-attachments)`
New mail | `gn` (for `go new`) | `nmap gn <plug>(iris-new-email)`
Previous page | `<Ctrl+b>` (for `page backward`) | `nmap <c-b> <plug>(iris-prev-page-emails)`
Next page | `<Ctrl+f>` (for `page forward`) | `nmap <c-f> <plug>(iris-next-page-emails)`
Change folder | `gf` (for `go folder`) | `nmap gf <plug>(iris-change-folder)`

### Email text preview

Function | Default keybind | Override
--- | --- | ---
Reply | `gr` (for `go reply`) | `nmap gr <plug>(iris-reply-email)`
Reply all | `gR` (for `go reply all`) | `nmap gR <plug>(iris-reply-all-email)`
Forward | `gf` (for `go forward`) | `nmap gf <plug>(iris-forward-email)`

### Email composition

Iris is based on the builtin `mail.vim` filetype and syntax. An email should
contains a list of headers followed by the message:

```vim
To: mail@test.com
Subject: Welcome

Hello world!
```

Function | Default keybind | Override
--- | --- | ---
Save draft | `:w` |
Send | `gs` (for `go send`) | `nmap gs <plug>(iris-send-email)`

### Folders

By default, Iris will display a basic prompt to select your folders:

```vim
:IrisFolder
```

Iris supports those fuzzy finders:

  - [fzf](https://github.com/junegunn/fzf) ([fzf.vim](https://github.com/junegunn/fzf.vim) need to be installed as well)

*Note: Iris will use the first fuzzy finder available automatically.*

### Flags

Flags appears in the first column of the [email list](#email-list) view. There
is 5 different flags:

  - `N` if it's a new email
  - `R` if it has been replied
  - `F` if it has been flagged
  - `D` if it's a draft
  - `@` if it contains an attachment

### Contacts

In order to autocomplete addresses, Iris keeps a `.contacts` file that contains
emails of your contacts. It's updated each time you send a new email (only the
`To` header is used). You can extract existing addresses from all your emails:

```vim
:IrisExtractContacts
```

*Note: the completion may need to be triggered manually via `<C-x><C-u>`, see
`:h i_CTRL-X_CTRL-U`.*

## Contributing

Git commit messages follow the [Angular
Convention](https://gist.github.com/stephenparish/9941e89d80e2bc58a153), but
contain only a subject.

  > Use imperative, present tense: “change” not “changed” nor
  > “changes”<br>Don't capitalize first letter<br>No dot (.) at the end

Code should be as clean as possible, variables and functions use the snake case
convention. A line should never contain more than `80` characters.

Tests should be added for each new functionality. Be sure to run tests before
proposing a pull request.

## Credits

  - [Neomutt](https://neomutt.org/)
  - [Alpine](http://alpine.x10host.com/alpine/alpine-info/)
  - [IMAPClient](https://github.com/mjs/imapclient)
    - Class [IMAPClient](https://github.com/mjs/imapclient/blob/580dc6781b5bf9d4f2a1a74b5d4168ef9b842b87/imapclient/imapclient.py#L162)
    - [Example](https://github.com/mjs/imapclient/blob/master/examples/example.py)
    - [Idle example](https://github.com/mjs/imapclient/blob/master/examples/idle_example.py)
  - IMAP [RFC3501](https://tools.ietf.org/html/rfc3501)
    - [Flags](https://tools.ietf.org/html/rfc3501#section-2.3.2)
    - [Status](https://tools.ietf.org/html/rfc3501#section-6.3.10)
    - [Search](https://tools.ietf.org/html/rfc3501#section-6.4.4)
    - [Fetch](https://tools.ietf.org/html/rfc3501#section-7.4.2)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Fixed

- Disconnection for inactivity [#25]

### Added

- fzf support for selecting folders [#29]

## [1.0.0] - 2020-06-09

### Added

- `In-Reply-To` header for threads [#7]
- Options to toggle idle mode [#11]
- Customizable keybinds [#18]
- Pagination [#12]
- Download attachments [#13]

### Changed

- Prevent asking passwords on startup via GPG [#2]
- Use vim builtin mail syntax `syntax/mail.vim` [#8]
- Split idle mode into its own subprocess [#9]
- Improve table column size and position [#21]

### Fixed

- Outlook issue [#3]
- Opening large mailbox [#4]
- Make use of `email.header.decode_header` to parse correctly headers [#6]
- Use `Reply-To` header to reply (if exists), otherwise `From` [#19]

### Security

- Hide passwords from logs [#20]

[unreleased]: https://github.com/soywod/iris.vim/compare/v1.0.0...HEAD
[1.0.0]: https://github.com/soywod/iris.vim/releases/tag/v1.0.0

[#2]: https://github.com/soywod/iris.vim/issues/2
[#3]: https://github.com/soywod/iris.vim/issues/3
[#4]: https://github.com/soywod/iris.vim/issues/4
[#6]: https://github.com/soywod/iris.vim/issues/6
[#7]: https://github.com/soywod/iris.vim/issues/7
[#8]: https://github.com/soywod/iris.vim/issues/8
[#9]: https://github.com/soywod/iris.vim/issues/9
[#11]: https://github.com/soywod/iris.vim/issues/11
[#12]: https://github.com/soywod/iris.vim/issues/12
[#13]: https://github.com/soywod/iris.vim/issues/13
[#18]: https://github.com/soywod/iris.vim/issues/18
[#19]: https://github.com/soywod/iris.vim/issues/19
[#20]: https://github.com/soywod/iris.vim/issues/20
[#21]: https://github.com/soywod/iris.vim/issues/21
[#25]: https://github.com/soywod/iris.vim/issues/25
[#29]: https://github.com/soywod/iris.vim/issues/29


api.py
#!/usr/bin/env python3

import json
import logging
import os
import quopri
import re
import smtplib
import subprocess
import sys
import threading

from base64 import b64decode
from email import policy
from email.header import Header, decode_header
from email.mime.text import MIMEText
from email.parser import BytesParser, BytesHeaderParser
from email.utils import formataddr, formatdate, make_msgid
from imapclient.imapclient import IMAPClient

log_filename = os.path.join(os.path.dirname(os.path.realpath(__file__)), "iris-api.log")
logging.basicConfig(filename=log_filename, format="[%(asctime)s] %(message)s", level=logging.INFO, datefmt="%Y-%m-%d %H:%M:%S")

imap_client = None
imap_host = imap_port = imap_login = imap_passwd = None
smtp_host = smtp_port = smtp_login = smtp_passwd = None

no_reply_pattern = r"^.*no[\-_ t]*reply"

def get_contacts():
    contacts = set()
    fetch = imap_client.fetch("1:*", ["ENVELOPE"])

    for [_, data] in fetch.items():
        envelope = data[b"ENVELOPE"]
        contacts = contacts.union(decode_contacts(envelope.to))

    return list(contacts)

def get_emails(last_seq, chunk_size):
    global imap_client

    emails = []
    if last_seq == 0:
        return emails

    ids = "%d:%d" % (last_seq, last_seq - chunk_size) if (last_seq > chunk_size) else "%d:%d" % (last_seq, 1)
    fetch = imap_client.fetch(ids, ["FLAGS", "ENVELOPE", "INTERNALDATE", "BODY.PEEK[HEADER]", "BODYSTRUCTURE", "FLAGS"])

    for [uid, data] in fetch.items():
        header = BytesHeaderParser(policy=policy.default).parsebytes(data[b"BODY[HEADER]"])
        envelope = data[b"ENVELOPE"]
        logging.info(data[b"BODYSTRUCTURE"])
        struct = data[b"BODYSTRUCTURE"][0] if isinstance(data[b"BODYSTRUCTURE"][0], list) else []
        has_attachment = len([mime[0] for mime in struct if mime[0] and mime[0] not in [b"text", b"multipart"]]) > 0
        subject = decode_byte(envelope.subject)
        from_ = envelope.from_[0]
        from_ = "@".join([decode_byte(from_.mailbox), decode_byte(from_.host)])
        to = envelope.to[0]
        to = "@".join([decode_byte(to.mailbox), decode_byte(to.host)])
        date_ = data[b"INTERNALDATE"].strftime("%d/%m/%y, %Hh%M")

        email = dict()
        email["id"] = uid
        email["subject"] = subject
        email["from"] = from_
        email["to"] = to
        email["date"] = date_
        email["flags"] = get_flags_str(data[b"FLAGS"], has_attachment)
        email["message-id"] = envelope.message_id.decode()
        email["reply-to"] = header["Reply-To"] if "Reply-To" in header else None

        emails.insert(0, email)

    return emails

def get_email(id, format):
    global imap_client

    fetch = imap_client.fetch([id], ["BODY[]"])
    content = get_email_content(id, fetch.popitem()[1][b"BODY[]"])

    return content[format]

def get_flags_str(flags, has_attachment):
    flags_str = ""

    flags_str += "N" if not b"\\Seen" in flags else " "
    flags_str += "R" if b"\\Answered" in flags else " "
    flags_str += "F" if b"\\Flagged" in flags else " "
    flags_str += "D" if b"\\Draft" in flags else " "
    flags_str += "@" if has_attachment else " "

    return flags_str
    
def download_attachments(dir, uid, data):
    attachments = []
    email = BytesParser(policy=policy.default).parsebytes(data)

    for part in email.walk():
        if part.is_attachment():
            attachment_name = part.get_filename()
            attachment = open(os.path.expanduser(os.path.join(dir, attachment_name)), "wb")
            attachment.write(part.get_payload(decode=True))
            attachment.close()
            attachments.append(attachment_name)

    return attachments

def get_email_content(uid, data):
    content = dict(text=None, html=None)
    email = BytesParser(policy=policy.default).parsebytes(data)

    for part in email.walk():
        if part.is_multipart():
            continue

        if part.get_content_type() == "text/plain":
            content["text"] = read_text(part)
            continue

        if part.get_content_type() == "text/html":
            content["html"] = read_html(part, uid)
            continue

    if content["html"] and not content["text"]:
        tmp = open(content["html"], "r")
        content["text"] = tmp.read()
        tmp.close()

    return content

def read_text(part):
    payload = part.get_payload(decode=True)
    return payload.decode(part.get_charset() or part.get_content_charset() or "utf-8")

def read_html(part, uid):
    payload = read_text(part)
    preview = write_preview(payload.encode(), uid)

    return preview

def write_preview(payload, uid, subtype="html"):
    preview = "/tmp/preview-%d.%s" % (uid, subtype)

    if not os.path.exists(preview):
        tmp = open(preview, "wb")
        tmp.write(payload)
        tmp.close()

    return preview

def decode_byte(byte):
    decode_list = decode_header(byte.decode())

    def _decode_byte(byte_or_str, encoding):
        return byte_or_str.decode(encoding or "utf-8") if type(byte_or_str) is bytes else byte_or_str

    return "".join([_decode_byte(val, encoding) for val, encoding in decode_list])

def decode_contacts(contacts):
    return list(filter(None.__ne__, [decode_contact(c) for c in contacts or []]))

def decode_contact(contact):
    if not contact.mailbox or not contact.host: return None

    mailbox = decode_byte(contact.mailbox)
    if re.match(no_reply_pattern, mailbox): return None

    host = decode_byte(contact.host)
    if re.match(no_reply_pattern, host): return None

    return "@".join([mailbox, host]).lower()

class PreventLogout(threading.Thread):
    def __init__(self):
        self.event = threading.Event()
        super(PreventLogout, self).__init__()
        self.start()

    def run(self):
        global imap_client
        while not self.event.wait(60):
            logging.info("NOOP")
            imap_client.noop()

while True:
    request_raw = sys.stdin.readline()

    try: request = json.loads(request_raw.rstrip())
    except: continue

    logging.info("Receive: " + str({key: request[key] for key in request if key not in ["imap-passwd", "smtp-passwd"]}))

    if request["type"] == "login":
        try:
            imap_host = request["imap-host"]
            imap_port = request["imap-port"]
            imap_login = request["imap-login"]
            imap_passwd = request["imap-passwd"]

            smtp_host = request["smtp-host"]
            smtp_port = request["smtp-port"]
            smtp_login = request["smtp-login"]
            smtp_passwd = request["smtp-passwd"]

            imap_client = IMAPClient(host=imap_host, port=imap_port)
            imap_client.login(imap_login, imap_passwd)
            PreventLogout()

            folders = list(map(lambda folder: folder[2], imap_client.list_folders()))
            response = dict(success=True, type="login", folders=folders)
        except Exception as error:
            response = dict(success=False, type="login", error=str(error))

    elif request["type"] == "fetch-emails":
        try:
            emails = get_emails(request["seq"], request["chunk-size"])
            response = dict(success=True, type="fetch-emails", emails=emails)
        except Exception as error:
            response = dict(success=False, type="fetch-emails", error=str(error))

    elif request["type"] == "fetch-email":
        try:
            email = get_email(request["id"], request["format"])
            response = dict(success=True, type="fetch-email", email=email, format=request["format"])
        except Exception as error:
            response = dict(success=False, type="fetch-email", error=str(error))

    elif request["type"] == "download-attachments":
        try:
            fetch = imap_client.fetch([request["id"]], ["BODY[]"])
            attachments = download_attachments(request["dir"], request["id"], fetch.popitem()[1][b"BODY[]"])
            response = dict(success=True, type="download-attachments", attachments=attachments)
        except Exception as error:
            response = dict(success=False, type="download-attachments", error=str(error))

    elif request["type"] == "select-folder":
        try:
            folder = request["folder"]
            seq = imap_client.select_folder(folder)[b"UIDNEXT"]
            emails = get_emails(seq, request["chunk-size"])
            is_folder_selected = True
            response = dict(success=True, type="select-folder", folder=folder, seq=seq, emails=emails)
        except Exception as error:
            response = dict(success=False, type="select-folder", error=str(error))

    elif request["type"] == "send-email":
        try:
            message = MIMEText(request["message"])
            for key, val in request["headers"].items(): message[key] = val
            message["From"] = formataddr((request["from"]["name"], request["from"]["email"]))
            message["Message-Id"] = make_msgid()

            smtp = smtplib.SMTP(host=smtp_host, port=smtp_port)
            smtp.starttls()
            smtp.login(smtp_login, smtp_passwd)
            smtp.send_message(message)
            smtp.quit()

            imap_client.append("Sent", message.as_string())

            contacts_file = open(os.path.dirname(sys.argv[0]) + "/.contacts", "a")
            contacts_file.write(request["headers"]["To"] + "\n")
            contacts_file.close()

            response = dict(success=True, type="send-email")
        except Exception as error:
            response = dict(success=False, type="send-email", error=str(error))

    elif request["type"] == "extract-contacts":
        try:
            contacts = get_contacts()
            contacts_file = open(os.path.dirname(sys.argv[0]) + "/.contacts", "w+")
            for contact in contacts: contacts_file.write(contact + "\n")
            contacts_file.close()

            response = dict(success=True, type="extract-contacts")
        except Exception as error:
            response = dict(success=False, type="extract-contacts", error=str(error))

    json_response = json.dumps(response)
    logging.info("Send: " + str(json_response))
    sys.stdout.write(json_response + "\n")
    sys.stdout.flush()


idle.py
#!/usr/bin/env python3

import json
import logging
import os
import subprocess
import sys

from imapclient.imapclient import IMAPClient

log_filename = os.path.join(os.path.dirname(os.path.realpath(__file__)), "iris-idle.log")
logging.basicConfig(filename=log_filename, format="[%(asctime)s] %(message)s", level=logging.INFO, datefmt="%Y-%m-%d %H:%M:%S")

request_raw = sys.stdin.readline()
request = json.loads(request_raw.rstrip())
logging.info({key: request[key] for key in request if key not in ["imap-passwd"]})

host = request["imap-host"]
port = request["imap-port"]
login = request["imap-login"]
passwd = request["imap-passwd"]
idle_timeout = request["idle-timeout"]

imap = IMAPClient(host, port)
imap.login(login, passwd)
imap.select_folder("INBOX")
imap.idle()

def notify_new_mail():
    title = "Iris"
    msg = "New mail available!"

    if sys.platform == "darwin":
        cmd = ["terminal-notifier", "-title", title, "-message", msg]
    else:
        cmd = ["notify-send", title, msg]

    logging.info("Notify: " + " ".join(cmd))
    subprocess.Popen(cmd)

while True:
    idle_res = imap.idle_check(timeout=idle_timeout)
    logging.info("Receive: " + str(idle_res))
    for res in idle_res:
        if res[1] == b"EXISTS": notify_new_mail()


iris.py
#!/usr/bin/env python3

import quopri
import re
import urllib
import vim

from base64 import b64decode
from datetime import date, datetime
from functools import reduce
from imapclient.imapclient import IMAPClient

def imap_connect():
    host = vim.vars['iris_host']
    email = vim.vars['iris_email']
    password = vim.eval('password')

    client = IMAPClient(host=host)
    client.login(email, password)
    client.select_folder('INBOX', readonly=True)

    sort = client.search(['NOT', 'DELETED', 'SINCE', date(2018, 9, 1)])
    response = client.fetch(sort, ['BODY.PEEK[HEADER]', 'INTERNALDATE'])
    client.logout()

    messages = []

    for [id, data] in response.items():
        headers = {}
        matches = re.findall('(.*?): (.*?)\r\n', data[b'BODY[HEADER]'].decode())

        for match in matches:
            headers[match[0]] = match[1]

        message = dict()
        message['id'] = str(id)
        message['subject'] = decode(headers['Subject'])
        message['from'] = decode(headers['From'])
        message['date'] = data[b'INTERNALDATE'].strftime('%d/%m/%y, %Hh%M')
        message['flags'] = '!'

        messages.append(message)

    return messages

def decode(string):
    match = re.match('^=\?(.*?)\?(.*?)\?(.*?)\?=$', string) 
    if (match == None): return string

    string_decoded = match.group(3)

    if (match.group(2).upper() == 'B'):
        string_decoded = b64decode(string_decoded)

    return quopri.decodestring(string_decoded).decode(match.group(1))


server.py
#!/usr/bin/env python3

import json
import logging
import os
import quopri
import re
import smtplib
import subprocess
import sys
import threading

from base64 import b64decode
from email import policy
from email.header import Header, decode_header
from email.mime.text import MIMEText
from email.parser import BytesParser
from email.utils import formataddr
from email.utils import formatdate
from email.utils import make_msgid
from imapclient.imapclient import IMAPClient

logging.basicConfig(filename="/tmp/iris.log", level=logging.INFO)

imap_client = None
imap_host = imap_port = imap_login = imap_passwd = None
smtp_host = smtp_port = smtp_login = smtp_passwd = None

idle_thread = None
is_idle = is_folder_selected = False

def notify_new_mail():
    title = "Iris"
    msg = "New mail available!"

    if sys.platform == "darwin":
        cmd = ["terminal-notifier", "-title", title, "-message", msg]
    else:
        cmd = ["notify-send", title, msg]

    subprocess.Popen(cmd)

def handle_idle():
    while True:
        global imap_client, is_idle

        if not is_idle: break
        idle_res = imap_client.idle_check(timeout=15)

        if idle_res: logging.info("idle received: " + str(idle_res))

        for res in idle_res:
            if res[1] == b"EXISTS": notify_new_mail()

def get_emails(last_seq):
    global imap_client

    emails = []
    if last_seq == 0:
        return emails

    chunk_size = 50
    ids = "%d:%d" % (last_seq, last_seq - chunk_size) if (last_seq > chunk_size) else "%d:%d" % (last_seq, 1)
    fetch = imap_client.fetch(ids, ["ENVELOPE", "INTERNALDATE", "FLAGS"])

    for [uid, data] in fetch.items():
        envelope = data[b"ENVELOPE"]
        subject = decode_byte(envelope.subject)
        logging.info(subject)
        from_ = envelope.from_[0]
        from_ = "@".join([decode_byte(from_.mailbox), decode_byte(from_.host)])
        to = envelope.to[0]
        to = "@".join([decode_byte(to.mailbox), decode_byte(to.host)])
        date_ = data[b"INTERNALDATE"].strftime("%d/%m/%y, %Hh%M")

        email = dict()
        email["id"] = uid
        email["subject"] = subject
        email["from"] = from_
        email["to"] = to
        email["date"] = date_
        email["flags"] = get_flags_str(data[b"FLAGS"])

        emails.insert(0, email)

    return emails

def get_email(id, format):
    global imap_client

    fetch = imap_client.fetch([id], ["BODY[]"])
    content = get_email_content(id, fetch.popitem()[1][b"BODY[]"])

    return content[format]

def get_flags_str(flags):
    flags_str = ""

    flags_str += "N" if not b"\\Seen" in flags else " "
    flags_str += "@" if b"\\Answered" in flags else " "
    flags_str += "!" if b"\\Flagged" in flags else " "
    flags_str += "#" if b"\\Draft" in flags else " "

    return flags_str
    
def get_email_content(uid, data):
    content = dict(text=None, html=None, attachments=[])
    email = BytesParser(policy=policy.default).parsebytes(data)

    for part in email.walk():
        if part.is_multipart():
            continue

        if part.is_attachment():
            content["attachments"].append(read_attachment(part, uid))
            continue

        if part.get_content_type() == "text/plain":
            content["text"] = read_text(part)
            continue

        if part.get_content_type() == "text/html":
            content["html"] = read_html(part, uid)
            continue

    if content["html"] and not content["text"]:
        tmp = open(content["html"], "r")
        content["text"] = tmp.read()
        tmp.close()

    return content

def read_text(part):
    payload = part.get_payload(decode=True)
    try: return quopri.decodestring(payload).decode()
    except: pass
    try: return payload.decode()
    except: return part.get_payload()

def read_html(part, uid):
    payload = read_text(part)
    preview = write_preview(payload.encode(), uid)

    return preview

def read_attachment(part, uid):
    payload = part.get_payload(decode=True)
    preview = write_preview(payload, uid, part.get_content_subtype())

    return preview

def write_preview(payload, uid, subtype="html"):
    preview = "/tmp/preview-%d.%s" % (uid, subtype)

    if not os.path.exists(preview):
        tmp = open(preview, "wb")
        tmp.write(payload)
        tmp.close()

    return preview

def decode_byte(byte):
    decode_list = decode_header(byte.decode())

    def _decode_byte(byte_or_str, encoding):
        return byte_or_str.decode(encoding or "utf-8") if type(byte_or_str) is bytes else byte_or_str

    return "".join([_decode_byte(val, encoding) for val, encoding in decode_list])

while True:
    request_raw = sys.stdin.readline()
    logging.info(request_raw.rstrip())

    try: request = json.loads(request_raw.rstrip())
    except: continue

    if request["type"] == "login":
        try:
            imap_host = request["imap-host"]
            imap_port = request["imap-port"]
            imap_login = request["imap-login"]
            imap_passwd = request["imap-password"]

            smtp_host = request["smtp-host"]
            smtp_port = request["smtp-port"]
            smtp_login = request["smtp-login"]
            smtp_passwd = request["smtp-password"]

            imap_client = IMAPClient(host=imap_host, port=imap_port)
            imap_client.login(imap_login, imap_passwd)

            folders = list(map(lambda folder: folder[2], imap_client.list_folders()))

            response = dict(success=True, type="login", folders=folders)
        except Exception as error:
            response = dict(success=False, type="login", error=str(error))

    elif request["type"] == "start-idle" and not is_idle:
        try:
            is_idle = True
            imap_client.idle()
            idle_thread = threading.Thread(target=handle_idle)
            idle_thread.setDaemon(True)
            idle_thread.start()
            response = dict(success=True, type="start-idle")
        except Exception as error:
            response = dict(success=False, type="start-idle")

    elif request["type"] == "stop-idle" and is_idle:
        try:
            logging.info('stop idle')
            is_idle = False
            idle_thread.join()
            imap_client.idle_done()
            response = dict(success=True, type="stop-idle")
        except Exception as error:
            response = dict(success=False, type="stop-idle")

    elif request["type"] == "fetch-emails":
        try:
            emails = get_emails(request["seq"])
            response = dict(success=True, type="fetch-emails", emails=emails)
        except Exception as error:
            response = dict(success=False, type="fetch-emails", error=str(error))

    elif request["type"] == "fetch-email":
        try:
            email = get_email(request["id"], request["format"])
            response = dict(success=True, type="fetch-email", email=email, format=request["format"])
        except Exception as error:
            response = dict(success=False, type="fetch-email", error=str(error))

    elif request["type"] == "select-folder":
        try:
            folder = request["folder"]
            seq = imap_client.select_folder(folder)[b"UIDNEXT"]
            emails = get_emails(seq)
            is_folder_selected = True
            response = dict(success=True, type="select-folder", folder=folder, seq=seq, emails=emails)
        except Exception as error:
            response = dict(success=False, type="select-folder", error=str(error))

    elif request["type"] == "send-email":
        try:
            from_name = request["headers"]["from-name"]
            from_email = request["headers"]["from-email"]

            message = MIMEText(request["message"])
            message["From"] = formataddr((from_name, from_email))
            message["To"] = request["headers"]["to"]
            message["Subject"] = Header(request["headers"]["subject"])
            message["Date"] = formatdate(localtime=True)
            message["Message-Id"] = make_msgid()

            if "cc" in request: message["CC"] = request["headers"]["cc"]
            if "bcc" in request: message["BCC"] = request["headers"]["bcc"]

            logging.info(message)
            smtp = smtplib.SMTP(host=smtp_host, port=smtp_port)
            smtp.starttls()
            smtp.login(smtp_login, smtp_passwd)
            smtp.send_message(message)
            smtp.quit()

            imap_client.append("Sent", message.as_string())

            response = dict(success=True, type="send-email")
        except Exception as error:
            response = dict(success=False, type="send-email", error=str(error))

    json_response = json.dumps(response)
    logging.info(json_response)
    sys.stdout.write(json_response + "\n")
    sys.stdout.flush()
